ICSharpCode.Decompiler/IL/Transforms/CachedDelegateInitialization.cs

```diff
@@ -236,5 +240,48 @@ bool CachedDelegateInitializationVB(IfInstruction inst)
 			inst.ReplaceWith(new StLoc(s, delegateConstruction));
 			return true;
 		}
+
+		/// <summary>
+		/// if (comp.o(ldobj delegateType(ldflda CachedAnonMethodDelegate(ldloc closure)) != ldnull)) Block {
+		/// 	stloc s(ldobj delegateType(ldflda CachedAnonMethodDelegate(ldloc closure)))
+		/// } else Block {
+		/// 	stloc s(stobj delegateType(ldflda CachedAnonMethodDelegate(ldloc closure), DelegateConstruction))
+		/// }
+		/// =>
+		///	stloc s(DelegateConstruction)
+		/// </summary>
+		bool CachedDelegateInitializationVBWithClosure(IfInstruction inst)
+		{
+			if (!(inst.TrueInst is Block trueInst && inst.FalseInst is Block falseInst))
+				return false;
+			if (trueInst.Instructions.Count != 1 || falseInst.Instructions.Count != 1)
+				return false;
+			if (!(trueInst.Instructions[0].MatchStLoc(out var s, out var trueInitValue)
+				  && falseInst.Instructions[0].MatchStLoc(s, out var falseInitValue)))
+			{
+				return false;
+			}
+			if (s.Kind != VariableKind.StackSlot || s.StoreCount != 2 || s.LoadCount != 1)
+				return false;
+			if (!(falseInitValue is StObj stobj) || !(trueInitValue is LdObj ldobj))
+				return false;
+			if (!(stobj.Value is NewObj delegateConstruction))
+				return false;
+			if (!stobj.Target.MatchLdFlda(out var target1, out var field1)
+				|| !ldobj.Target.MatchLdFlda(out var target2, out var field2)
+				|| !field1.Equals(field2) || !target1.Match(target2).Success)
+			{
+				return false;
+			}
+			if (!inst.Condition.MatchCompNotEqualsNull(out ILInstruction left))
```

Comment:
Shouldn't we account for the == null case as well?

Reply:
I based this new code on `CachedDelegateInitializationVB` which also does not have any special handling for inverted if conditions. If it is deemed necessary to add support for inverted if conditions, I can do so for both the new `CachedDelegateInitializationVBWithClosure` and `CachedDelegateInitializationVB` methods.

Reply:
It's also worth mentioning that testing this pattern matching in an automated fashion for various VB compiler versions is impossible using VBPretty tests as ILSpy does not support VB.NET Closure types (the equivalent of C# display classes) which leads to ugly decompiled code.